specification.md

```diff
@@ -1357,44 +1359,86 @@ For more information about workflow expressions, reference the [Workflow Express
-in a workflow definition. If a different expression language needs to be used, make sure to set the workflow 
-`expressionLang` property to identify it to runtime implementations.
+Expressions are evaluated during execution of the workflow and can be used for the following purposes:
+- Transform state data
+  - upon entering and leaving a state, using [State Data Filter](#State-data-filters) `input` and `output`
+  - entering and leaving an action, using [Action Data Filter](#Action-data-filters) `fromStateData`, `result`
+  - when consuming an event, using [Event Data Filter](#Event-data-filters) `data`
+- Select a path in the workflow data
+  - to set the result of an action to, using [Action Data Filter](#Action-data-filters) `toStateData`
+  - to set the result of consuming an event to, using [Event Data Filter](#Event-data-filters) `toStateData`
+- Define functions that can be executed as an action [using functions for expression evaluation](#Using-Functions-For-Expression-Evaluation)
+- Define any state parameter of the workflow definition.
+  Basically any leaf parameter in states can be set by evaluating an expression against the workflow state data at runtime. However, structures like the `actions` array in an [operation state](#Operation-state) can not be generated at runtime, i.e. the workflow definition has to conform to the JSON specification and contain all required elements. The JSON specification explicitly allows expressions where applicable.
+
+A workflow expression can be defined where it is applied as an _expression string_ 
+or it can be defined as an [expression function definition](#Using-Functions-For-Expression-Evaluation) 
+and then applied using an _expression reference string_.
+
+An _expression string_ starts with the `${` token and ends on `}`. All text 
+between the curly brackets is evaluated using the workflow's specified 
+[`expressionLang`](#Workflow-Definition-Structure).
+Note, that without the closing bracket at the very end of the string, it won't 
+be evaluated as an expression.
+
+A workflow expression can also be defined in an [expression function definition](#Using-Functions-For-Expression-Evaluation) and then applied by reference.
+To reference a expression function, an _expression reference string_ is used that 
+starts with the `$fn{` token and ends with `}`.
+The _expression reference string_ has the format `$fn{<function name>}`, where 
+`<function name>` has to match the user-specified `name` of the defined 
+expression function of `type` `"expression"`.
+
+On the rare occasion that one needs to create a string that literally looks like 
+an _expression string_ or an _expression reference string_, the dollar sign 
+should be repeated. Only if the remainder matches a workflow expression markup,
+the runtime reduces the double dollar signs `$$` to a single dollar sign `$` and 
+refrains from evaluating it as a workflow expression.
```

Comment:
Thanks, personally I wouldn't replace any occurence of a `$` at the start of an expression-capable string. Only the ones that could actually be interpreted as an expression (`${...}` or `$fn{...}`). I agree any markup like `$xyz{ }` could be included (e.g. regex `^\$.+\{.*\}`) that would allow for more than just the function reference. I was hoping to get a discussion on $fn{}. The existing markup `fn:` causes confusion for jq and maybe causes problems in other expression languages, too.